dvc/command/gc.py

```diff
@@ -39,9 +48,9 @@ def run(self):
             return 1
 
         self.repo.gc(
-            all_branches=self.args.all_branches,
-            all_tags=self.args.all_tags,
-            all_commits=self.args.all_commits,
+            all_branches="branches" in self.args.remove,
+            all_tags="tags" in self.args.remove,
+            all_commits="commits" in self.args.remove,
```

Comment:
Let's pass `remove` to the `repo.gc` and let it deal with it. We need to keep API and CLI symmetric here, no reason to change one and keep the other.

Reply:
Agree about the symmetricity. But, I find it better to use args rather than a collection of *filters*, kinda like buttons which you can turn on or off to be more aggressive/effective. 

I don't have strong opinions though.

Reply:
@skshetry well, but why not? :) Think about them as being used in future to remove specific references. E.g. `remove=mytag`, which you will have to pass as `remove=["mytag"]` to API anyways, so why bother with additional flags for `repo.gc`, if you can `repo.gc(remove=["commits", "tags", "branches"])` right away.

Reply:
Yes, it makes sense from *future*, as it allows a lot of flexibility. I was just saying, `all_commits=True, all_branches=True` are better interface than `remove={"commits", "branches"}`, unless there are too many arguments to pass (not really flexible).

I was looking from user interface-side. I'll fix.

Reply:
`all_commits=True`, ... is unsafe, this is why we switch to requiring user explicitly state what he/she wants to remove. If we think of `Repo` methods as somewhat potentially public then we should mirror cli interface in `Repo.gc()` interface. It might be translated there to `all_commits` and friends.

Reply:
I see three downsides to this, on top of @efiops point:
- it's very confusing that `all_branches` in `Repo.gc()` is somewhat but not entirely reverse to `all_branches` in `Repo.used_cache()/brancher()`
- three params instead of one combining them, I was looking into reducing the number of params in our functions and combining them is one of the ways
- we try to keep `Repo` methods API close to cli

Reply:
I assumed this was only related to CLI at first. I only realized very late that `all_commits=False` is not `--remove=commits`, but `--remove=commits,branches,tags`. Would it be good to make adjustments to the `Repo.brancher()` to exclude HEAD revisions from branches and  tags if `all_commits=True`  unless `all_tags` and/or `all_branches` is specified?

It'd be easier, I guess there, but, I'd like to handle it on `Repo.gc()` itself.

Reply:
@skshetry I would probably consider adding `exclude` argument to the brancher, that you could then pass something like `exclude=(scm.list_branches() + scm.list_tags())`

Reply:
@skshetry I would consider introducing an `exclude` flag for brancher and using it something like this: `brancher(all_commits=True, exclude=[scm.list_branches()])` to exclude branch heads. Other approach might be to collect used cache with `all_commit`, then with `all_branches` and substitute them 🙂

Reply:
@efiop I don't like changing brancher. The logic in both brancher and `.used_cache()` should stay the same. There is no reason to translate cli/api to internal code.

Reply:
@efiop additionally all that new logic you propose is more complex than existing one and does not solve anything that is required by this PR/issue.

Reply:
@Suor Let's be constructive here. What do you suggest?

Reply:
@efiop wrote that below, but maybe to make it more clear:
- leave `.brancher()`, `.used_cache()` and any non-cli non Repo api-methods code unchanged.
- accept new options in `dvc gc`, raise some `DeprecatedError` on the old ones
- pass that to `Repo.gc()` as is
- translate `remove={...}` to `all_branches` and friends that `.brancher()` and `.used_cache()` understand in `Repo.gc()` implementation.

This way only `CmdGC` and `Repo.gc()` would be changed.

Reply:
> accept new options in dvc gc, raise some DeprecatedError on the old ones

No need, we are breaking the default behavior anyway, preserving old options would be hard and confusing. We need to drop them right away, without any DeprecationError's.

I think you are missing something here. There is no direct translation of current options into the new ones, so your proposition is impossible to implement without changing some of that. For example, `dvc gc --remove=branches` needs to collect all used cache for all references except for all branches, so it would translate to something like `git rev-list  --all --skip $ALL_BRANCHES`, which requires adjustments for `used_cache` and `brancher`.